### ISSUE TITLE
enhancement: responsive layout

### ISSUE THREAD

**Issue by jgravois** [asalant/freehub#32]
greetings! 👋 

first of all, **thank you**. i've been using freehub happily for many years, first at the [Sacramento Bike Kitchen](http://www.sacbikekitchen.org/) and later at [BikeBBQ](https://www.bikebbq.com) in Redlands.

1) awhile back i helped bootstrap another nearby coop called the [San Bernardino Bicycle Hubitat](https://www.facebook.com/bicyclehubitatSB/). a volunteer who i'm no longer in touch with created an account on https://freehub.bikekitchen.org and used `hubitat` as the username and subdomain. i'm under the impression that it was never actually put to use. i understand if its not possible, but if there _is_ a way to associate that username with a new email and password, i'd be much obliged.

2) how can i make a financial contribution on behalf of our bike coop to help y'all keep the lights on?

3) how would y'all feel about a PR that makes the application easier to navigate on a mobile device?

**Comment by asalant** [comment]
Hi John! I just found your message buried in my inbox. I'll respond to you with details shortly.

**Comment by jgravois** [comment]
@asalant I ended up just creating another account instead of trying to recover the first one.

that said, I've got a staff member using a phone as the primary device for signing folks in so I'm still keen to PR some new CSS that makes the application easier to navigate on small screens if there's any chance of it actually landing.

**Comment by asalant** [comment]
Ugh. Sorry John. I truly intended to make time to dig in to that for you
but am on extended travel with young kids and it is shockingly hard to find
a couple hours in front of a computer. I should have just told you I wasn't
sure when I'd get to out.

That said, I'm back home in July and would love to merge a PR that adds
mobile support. Can you believe that Freehub was released before that was
such a common thing? And not that long ago!

**Comment by jgravois** [comment]
no worries at all. enjoy the time AFK.

i'll take a stab at setting up the development environment and laying out the application using css-grid sometime soon.

**Comment by fspc** [comment]
@jgravois , there is a Dockerfile at https://github.com/fspc/freehub that could help you setup a development environment. 

For simplicity,  supervisord runs mysql and freehub in the same container, although that could easily be accomplished in separate containers without supervisord.

**Comment by asalant** [comment]
Good luck! It is an _old_ Rails install. Someone else has done some work to
get the environment up to date. If you run into issues, I can put you in
touch.

**Comment by jgravois** [comment]
thanks for the pointer @fspc. i was able to set up a local development environment (with hardly any docker experience) by:

1. pulling down and running your handy container
2. `git clone`ing `asalant/freehub`
3. using [`onchange`](https://www.npmjs.com/package/onchange) to copy/paste the contents of the cloned repo _into_ the same location in the container each time i touch a file.

```
onchange '**/*' -- docker cp . freehub:/freehub
```

**Comment by fspc** [comment]
@jgravois, good to hear, and a cool use of onchange!

> thanks for the pointer @fspc. i was able to set up a local development environment (with hardly any docker experience) by:
>
> 3. using [`onchange`](https://www.npmjs.com/package/onchange) to copy/paste the contents of the cloned repo _into_ the same location in the container each time i touch a file.
> 
> ```
> onchange '**/*' -- docker cp . freehub:/freehub
> ```

**Comment by asalant** [comment]
Hi @jgravois! Just an FYI that I have some time to dedicate to Freehub and am working on a couple updates. 

First is purely backend - I've got to do some hosting and infrastructure changes by Jan 1 as requested by our provider Engine Yard who has been providing pro bono hosting.

But once that is done, responsive layouts to make it easy to use Freehub on phones is my top priority. Let me know if you've made any progress there or want to collaborate.

@fspc I'll be trying to work with your Docker config too.

**Comment by jgravois** [comment]
> I have some time to dedicate to Freehub and am working on a couple updates.

very cool!

after I got things up and running locally and played around with the .css a bit I promptly had a new baby 👶 and set the whole project aside.

its funny that you mention it though, because this thanksgiving break I made a promise to myself to pick this back up and chip away at it a little at a time so I am _definitely_ game to collaborate.

the first time around I thought I'd be clever (and helpful) to make a lot less extensive use of ids and use classnames instead.

The more I think about it though, the more I realized that I'd probably be better off leaving the backend alone (for now). to keep me honest on that front my new plan has been to just save static .html from the live site and start building up a new .css file one page at a time

- [ ] homepage
- [ ] /:org
- [ ] /:org/visits
- [ ] etc.

how would you feel about doing it that way? if so, I'll plan on saving some boilerplate in a new public repo of my own and inviting you as a collaborator.

**Comment by asalant** [comment]
Ah, babies. Welcome to parenthood! I've forgotten what it was like to have
free time before kids.

My first thoughts about how to proceed with a responsive layout are around
development and release strategy. We have quite a few people using Freehub
daily and there have been no changes in years. It might be helpful to
develop and release a responsive layout in parallel with the current
experience, then put up a notice for users to try it and give
feedback/report bugs, wait a couple weeks then release it to everyone. The
mechanism to do this could be a feature flag as a request parameter (e.g.
?responsive=true) that is detected by a controller that then uses a
different layout and view to render the page.

I'm a big fan of releasing code to production as it is worked on but in a
way that does not force users to use it until it is ready. It may be
especially helpful if we're working on it ad hoc.

Regarding working on the actual HTML/CSS, I'd say let's prioritize the
views by where a responsive layout will deliver the most value. Then work
on the first one until it is done before moving on to the next. Do you have
any opinions for view priority based on your use of Freehub?

My focus for this week is the system upgrades so I'm not sure if I'll be
getting to responsive this week. Great to hear I may get some help ;)

**Comment by jgravois** [comment]
> I'm a big fan of releasing code to production as it is worked on but in a
way that does not force users to use it until it is ready.

that sounds good to me, but it may get a bit sticky trying to load both the new .css and old .css simultaneously. another option would be to (temporarily) host http://responsive-freehub.bikekitchen.org temporarily and try to gather feedback from users for a few weeks before flipping the switch for real.

> let's prioritize the views by where a responsive layout will deliver the most value.

that makes a ton of sense, for us that would be the staff login and customer sign in pages. I'll start playing around with them first.

**Comment by jgravois** [comment]
done in #43 🎉🚀:shipit: 

I'll be able to collect some feedback from my San Bernardino Bicycle Hubitat user over the next few weeks.
